Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc

```diff
@@ -2539,390 +2819,469 @@ void HGVHistoProducerAlgo::tracksters_to_SimTracksters(const Histograms& histogr
-      if (!cp_considered_efficient && tstSharedEnergyFrac[cpId][tstId] >= minTSTSharedEneFracEfficiency_) {
+      if (!cp_considered_efficient && (tstSharedEnergyFrac >= minTSTSharedEneFracEfficiency_)) {
         cp_considered_efficient = true;
-        histograms.h_numEff_caloparticle_eta[count]->Fill(simTSFromCP[iSTS].barycenter().eta());
-        histograms.h_numEff_caloparticle_phi[count]->Fill(simTSFromCP[iSTS].barycenter().phi());
+        histograms.h_numEff_caloparticle_eta[i][count]->Fill(sts_eta);
+        histograms.h_numEff_caloparticle_phi[i][count]->Fill(sts_phi);
+        histograms.h_numEff_caloparticle_en[i][count]->Fill(sts_en);
+        histograms.h_numEff_caloparticle_pt[i][count]->Fill(sts_pt);
       }
-    }  //end of loop through Tracksters
-
-    auto is_assoc = [&](const auto& v) -> bool { return v < ScoreCutCPtoTSEffDup_; };
-
-    auto assocDup = std::count_if(std::begin(score3d[cpId]), std::end(score3d[cpId]), is_assoc);
-
-    if (assocDup > 0) {
-      histograms.h_num_caloparticle_eta[count]->Fill(simTSFromCP[iSTS].barycenter().eta());
-      histograms.h_num_caloparticle_phi[count]->Fill(simTSFromCP[iSTS].barycenter().phi());
-      auto best = std::min_element(std::begin(score3d[cpId]), std::end(score3d[cpId]));
-      auto bestTstId = std::distance(std::begin(score3d[cpId]), best);
-
-      histograms.h_sharedenergy_caloparticle2trackster_vs_eta[count]->Fill(
-          simTSFromCP[iSTS].barycenter().eta(), tracksters[bestTstId].raw_energy() / CPenergy);
-      histograms.h_sharedenergy_caloparticle2trackster_vs_phi[count]->Fill(
-          simTSFromCP[iSTS].barycenter().phi(), tracksters[bestTstId].raw_energy() / CPenergy);
-      LogDebug("HGCalValidator") << count << " " << simTSFromCP[iSTS].barycenter().eta() << " "
-                                 << simTSFromCP[iSTS].barycenter().phi() << " " << tracksters[bestTstId].raw_energy()
-                                 << " " << CPenergy << " " << (tracksters[bestTstId].raw_energy() / CPenergy) << " "
-                                 << tstSharedEnergyFrac[cpId][bestTstId] << '\n';
-      histograms.h_sharedenergy_caloparticle2trackster_assoc[count]->Fill(tstSharedEnergyFrac[cpId][bestTstId]);
-
-      if (assocDup >= 2) {
-        auto match = std::find_if(std::begin(score3d[cpId]), std::end(score3d[cpId]), is_assoc);
-        while (match != score3d[cpId].end()) {
-          tracksters_duplicate[std::distance(std::begin(score3d[cpId]), match)] = 1;
-          match = std::find_if(std::next(match), std::end(score3d[cpId]), is_assoc);
-        }
+
+      if (score3d_iSTS[tstId] < ScoreCutSTStoTSPurDup) {
+        if (tracksters_PurityDuplicate[tstId] < 1)
+          tracksters_PurityDuplicate[tstId]++;  // for Purity
+        if (cp_considered_pure)
+          tracksters_PurityDuplicate[tstId]++;  // for Duplicate
```

Comment:
Ciao @lecriste,
it is not immediate to grasp the meaning and the purpose of the variable `cp_considered_pure`.
Could you please clarify that and explain the rationale behind this code snippet?

Reply:
A `SimTrackster` is considered "pure" when matched to at least one `recoTrackster`.
A `recoTrackster` is considered "duplicated" if matched to a `SimTrackster` that has already another `recoTrackster` matched, i.e. already "pure".
The `cp` in `cp_considered_pure` does not refer only to CaloParticle, but to any `SimTrackster`.

Reply:
Ciao @lecriste thanks for the clarification. I'd use a more generic name since `cp_` in our code really means `CaloParticle`.
Also, I was wondering if you could simply remove all the `if` logic and the `cp_considered_pure` variable itself and simply increment the counter every time there's a match? What would be missing in this case?

Reply:
Without the `if` logic, when the same `recoTrackster` (`tstId`) matches more than one `SimTrackster` then all the matchings would increment the counter and mark the `recoTrackster` as duplicated.